### Pandas Lab - Finding, Querying, Creating Data

### Section 1: Selecting Data

Complete the following prompts, and compare your results with the answers to confirm that you did the operation properly.

**1).** Find the average age of all passengers on board.  The answer will be 29.7

In [10]:
import numpy as np
import pandas as pd

df = pd.read_csv(r'C:\Users\James.Kelly\DAT-06-24\class material\Unit 2\data\titanic.csv')


df.Age.mean()
df['Age'].mean()

29.69911764705882

first 100 rows )**2).** Find the median value of the 'Fare' and 'SibSp' columns.  The answer will be 14.45 and 0.00, respectively.

In [2]:
print(df.Fare.median())
print(df.SibSp.median())

df[['Fare','SibSp']][:100].median()

14.4542
0.0


Fare     15.675
SibSp     0.000
dtype: float64

**3).** Find the average age of all passengers on board.  The answer will be 15.675 and 0.00.

In [40]:
df[['Age']].mean()

Age    29.699118
dtype: float64

**4).** Using the .iloc command, grab the modal value of the last 4 columns in the dataset.  The result should be a 3x3 dataframe that has the values 1601, 8.05, B96 B98, and S in the first row.

In [3]:
df.iloc[:,-4:].mode()

Ticket  Fare        Cabin Embarked
0      1601  8.05      B96 B98        S
1    347082   NaN  C23 C25 C27      NaN
2  CA. 2343   NaN           G6      NaN

**5).** Using the .iloc command, grab the mean value of the first 250 rows of the first 3 columns in the dataset.  The answer should be:
    - Passengerid: 125.5
    - Survived: 0.344
    - Pclass: 2.416

In [48]:
df.iloc[:250,:3].mean()

PassengerId    125.500
Survived         0.344
Pclass           2.416
dtype: float64

### Section 2: Querying Data

**1).** How many females were on board the Titanic? Men? The answers should be 314, 577

In [4]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [71]:
print("number of women: " + str(df[df.Sex == 'female']['PassengerId'].count()))
print("number of men: " + str(df[df.Sex == 'male']['PassengerId'].count()))

number of women: 314
number of men: 577


**2).** What was the survival rate for females on the Titanic? The answer should be 0.74

In [80]:
print("survival rate for women: "+ str(df[df.Sex == 'female']['Survived'].mean()))

survival rate for women: 0.7420382165605095


**3).** What was the survival rate for Males? The answer should be 0.19

In [86]:
print("survival rate for men: "+ str(df[df.Sex == 'male']['Survived'].mean()))

survival rate for men: 0.18890814558058924


**4).** What was the survival rate for passengers in either Pclass 1 or Pclass 2?  The answer should be 0.55

In [89]:
print("survival rate for passengers in either Pclass 1 or Pclass 2: " + str(df[(df.Pclass == 1) | (df.Pclass == 2)]['Survived'].mean()))

survival rate for passengers in either Pclass 1 or Pclass 2: 0.5575


**5).** What was the survival rate if you were female and had at least 1 sibling on board? The answer should be 0.686.

In [90]:
np.round(df[(df.Sex == 'female') & (df.SibSp >= 1)]['Survived'].mean(), 2)

0.69

### Section 3: Creating New Data

**1).** Create a column called 'Is_Alone' that's either True or False, depending on whether or not the column 'Family_Size' is > 0 or not.

In [39]:
df['Family_Size'] = df.SibSp + df.Parch
df['Is_Alone'] = np.where(df['Family_Size'] > 0, True, False)

**2).** Create a column called 'Demographic' that breaks people up into the following categories:
 - Below 8 years old: 'Child'
 - 8 - 21 years old: 'Adolescent'
 - 22 - 55 years old: 'Adult'
 - 55+: 'Senior'
 
When you're finished, use the method .value_counts() to confirm each section has the following count values:
 - Child: 49
 - Adolescent: 154
 - Adult: 470
 - Senior: 40

In [6]:
df['Age'].max()

80.0

In [40]:
df['Demograhic'] =  pd.cut(df['Age'], bins=[df['Age'].min() - 1,7,21,55,df['Age'].max() +1], labels=['Child', 'Adolescent', 'Adult', 'Senior'])

In [41]:
df['Demograhic'].value_counts()

Adult         470
Adolescent    154
Child          50
Senior         40
Name: Demograhic, dtype: int64

**3).** Create a column called 'Gender_Status' that returns the following values:

 - 'F-High' if passenger is female and passenger class is 1.
 - 'F-Low' if passenger is female and passenger class is 2 or 3
 - 'M-High' if passenger is male and passenger class is 1
 - 'M-Low' if passenger is male and passenger class is 2 or 3
 
When you are finished, use .value_counts() to confirm you have the following count values for each of the following:
 - 'M-Low': 455
 - 'F-Low': 220
 - 'M-High': 122
 - 'F-High': 94

In [22]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  Demograhic  
0      0         A/5 21171   7.2500   NaN        S  Adolescent  
1      0          PC 17599  71.2833   C85        C       Adult  
2      0  STON/O2. 3101282   7.9250   NaN        S       Adult  
3      0            113803  53.1000  C123        S       Adult  
4      0            373450   8.0500   NaN        S       Adult

In [24]:
conditions = [
    (df['Sex'] == 'female') & (df['Pclass'] == 1),
    (df['Sex'] == 'female') & (df['Pclass'] > 1),
    (df['Sex'] == 'male') & (df['Pclass'] == 1),
    (df['Sex'] == 'male') & (df['Pclass'] > 1)
]

results = [
    'F-High',
    'F-Low',
    'M-High',
    'M-Low'
]

df['Gender_Status'] = np.select(conditions, results, 'Other')

df['Gender_Status'].value_counts()

M-Low     455
F-Low     220
M-High    122
F-High     94
Name: Gender_Status, dtype: int64

**4).** Using string methods, extract the *greeting* of the persons name on board.  Ie, if someone's name is 'Ms. Madame Bovary', create a column that contains the value 'Ms.' and nothing else.

**Hint:** Take a look at the split() method and see what it does if you're not sure where to go.

In [43]:
df['greeting'] = df['Name'].str.split().str[1]

In [58]:
df['greeting'].head(10)

0        Mr.
1       Mrs.
2      Miss.
3       Mrs.
4        Mr.
5        Mr.
6        Mr.
7    Master.
8       Mrs.
9       Mrs.
Name: greeting, dtype: object

In [66]:
df['greeting'].value_counts()

Mr.             502
Miss.           179
Mrs.            121
Master.          40
Dr.               7
Rev.              6
y                 4
Planke,           3
Impe,             3
Major.            2
Gordon,           2
Col.              2
Mlle.             2
the               1
Messemaeker,      1
Walle,            1
Don.              1
Capt.             1
Shawah,           1
Cruyssen,         1
Billiard,         1
Jonkheer.         1
Melkebeke,        1
Pelsmaeker,       1
der               1
Velde,            1
Mme.              1
Mulder,           1
Steen,            1
Ms.               1
Carlo,            1
Name: greeting, dtype: int64

In [61]:
df['Name'].unique

<bound method Series.unique of 0                                Braund, Mr. Owen Harris
1      Cumings, Mrs. John Bradley (Florence Briggs Th...
2                                 Heikkinen, Miss. Laina
3           Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                               Allen, Mr. William Henry
5                                       Moran, Mr. James
6                                McCarthy, Mr. Timothy J
7                         Palsson, Master. Gosta Leonard
8      Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)
9                    Nasser, Mrs. Nicholas (Adele Achem)
10                       Sandstrom, Miss. Marguerite Rut
11                              Bonnell, Miss. Elizabeth
12                        Saundercock, Mr. William Henry
13                           Andersson, Mr. Anders Johan
14                  Vestrom, Miss. Hulda Amanda Adolfina
15                      Hewlett, Mrs. (Mary D Kingcome) 
16                                  Rice, Master. Eugene
